### PR TITLE
T 09 item backend

### PR DESCRIPTION
**Pull request by Ali-Alkhars**
Hello everyone! T-09 (implementing the add, delete and edit buttons for choice board items) is done.

### These are the main things I added:
- add item widget
- edit item widget
- delete item alert dialogue
- A class for firebase manipulation functions (handling backend)
- Widget tests and Unit tests for all added classes
- Had to do lots of refactoring to make classes able to take mock firebase functions for testing

_**Note:**_ Manual testing is very tricky because loading data from the database is not yet implemented. If you would like to manually test:
- go to our firestore console 
- manually add three categories (with correct fields)
- Take the ids of these created categories
- Go to flutter: lib/data/choice_boards_data.dart
- Paste one id to each Draggable list (change the current data)
- Create items using the app (add an item) buttons
- Copy the ids of the items from the firestore console
- Paste them into the ids of the Draggable list items
- Now you should also be able to edit/delete items using the app buttons

### CODE REVIEW

File: lib/helpers/mock_firebase_authentication.dart [hunk 1]
```diff
@@ -0,0 +1,4 @@
+import 'package:firebase_auth/firebase_auth.dart';
+import 'package:mockito/mockito.dart';
+
+class MockFirebaseAuthentication extends Mock implements FirebaseAuth {}
```
Comment (AymenBerb) [inline lib/helpers/mock_firebase_authentication.dart:4]: What is this file for ?

File: lib/widgets/hero_dialog_route.dart [hunk 2]
```diff
@@ -0,0 +1,46 @@
+import 'package:flutter/material.dart';
+
+/// Custom [PageRoute] that creates an overlay dialog (popup effect).
+/// taken from: https://youtu.be/Bxs8Zy2O4wk
+
+class HeroDialogRoute<T> extends PageRoute<T> {
+  /// {@macro hero_dialog_route}
```
Comment (absirgue) [inline lib/widgets/hero_dialog_route.dart]: delete this comment

**Review by absirgue — APPROVED** [review]
Awesome work sir, clean and very well tested.

Just this tiny comment to delete and it's perfect.

**Review by Digital-Brendan — APPROVED** [review]
Everything looks great - nicely designed and commented.

If I was being picky, I would give two particular functions more concise names, specifically `handleSavingItemToFirestore` and `handleEditingItemInFirestore`.

How about `saveItemToFirestore` or just `saveItem`?

Generally though, great code!

**Review by AymenBerb — CHANGES_REQUESTED** [review]
Screenshot 2023-03-03 at 16 03 14

Can you make item's name as a more recgnizable text field ? and maybe reduce the size of the pop up. (it doesnt need to take the whole screen (maybe add paading left and right))

File: lib/helpers/mock_firebase_authentication.dart (hunk 1 above)
Comment (Ali-Alkhars) [reply lib/helpers/mock_firebase_authentication.dart:4]: I need it to mock FirebaseAuthentication

**Review by AymenBerb — APPROVED** [review]
well done !

**Pull request merged**